### PR TITLE
Add support for pbkdf2 password algorithm

### PR DESCRIPTION
**Pull request by BurntimeX**

### CODE REVIEW

File: wcfsetup/install/files/lib/system/user/authentication/password/algorithm/Pbkdf2.class.php [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+{
+    /**
+     * @inheritDoc
+     */
+    public function verify(
+        #[\SensitiveParameter]
+        string $password,
+        string $hash
+    ): bool {
+        $parts = \explode(':', $hash, 5);
```
Comment (dtdesign) [inline wcfsetup/install/files/lib/system/user/authentication/password/algorithm/Pbkdf2.class.php:26]:
```suggestion
        $parts = \explode(':', $hash, 5);
        if (\count($parts) !== 5) {
            return false;
        }
```
I thought about an `\assert()` but I’m not sure if that makes sense at all.

File: wcfsetup/install/files/lib/system/user/authentication/password/algorithm/Pbkdf2.class.php [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+        $hash = $parts[0];
+        $salt = $parts[1];
+        $algo = $parts[2];
+        $iterations = $parts[3];
+        $length = $parts[4];
```
Comment (dtdesign) [inline wcfsetup/install/files/lib/system/user/authentication/password/algorithm/Pbkdf2.class.php:31]:
```suggestion
        [$hash, $salt, $algo, $iterations, $length] = $parts;
```

File: wcfsetup/install/files/lib/system/user/authentication/password/algorithm/Pbkdf2.class.php [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+        #[\SensitiveParameter]
+        string $password
+    ): string {
+        $salt = Hex::encode(\random_bytes(20));
+        $algo = 'sha256';
+        $iterations = 600000;
+        $length = 32;
+        $hash = \hash_pbkdf2($algo, $password, $salt, $iterations, $length);
+
+        return implode(':', [$hash, $salt, $algo, $iterations, $length]);
```
Comment (dtdesign) [inline wcfsetup/install/files/lib/system/user/authentication/password/algorithm/Pbkdf2.class.php:49]:
```suggestion
        return \implode(':', [$hash, $salt, $algo, $iterations, $length]);
```

**Review by dtdesign — APPROVED** [review]

File: wcfsetup/install/files/lib/system/user/authentication/password/algorithm/Pbkdf2.class.php [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+     */
+    public function hash(
+        #[\SensitiveParameter]
+        string $password
+    ): string {
+        $salt = Hex::encode(\random_bytes(20));
+        $algo = 'sha256';
+        $iterations = 600000;
+        $length = 32;
+        $hash = \bin2hex(\hash_pbkdf2($algo, $password, $salt, $iterations, $length, true));
```
Comment (TimWolla) [inline wcfsetup/install/files/lib/system/user/authentication/password/algorithm/Pbkdf2.class.php:46]:
```suggestion
        $hash = Hex::encode(\hash_pbkdf2($algo, $password, $salt, $iterations, $length, true));
```

File: wcfsetup/install/files/lib/system/user/authentication/password/algorithm/Pbkdf2.class.php [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+        string $password,
+        string $hash
+    ): bool {
+        $parts = \explode(':', $hash, 5);
+        if (\count($parts) !== 5) {
+            return false;
+        }
+        [$hash, $salt, $algo, $iterations, $length] = $parts;
+
+        return \hash_equals($hash, \bin2hex(\hash_pbkdf2($algo, $password, $salt, $iterations, $length, true)));
```
Comment (TimWolla) [inline wcfsetup/install/files/lib/system/user/authentication/password/algorithm/Pbkdf2.class.php:32]:
```suggestion
        return \hash_equals($hash, Hex::encode(\hash_pbkdf2($algo, $password, $salt, $iterations, $length, true)));
```

**Pull request merged**